include/viz/sfml/robot_control.h

```diff
@@ -0,0 +1,130 @@
+#pragma once
+
+// BoB robotics includes
+#include "robots/type_traits.h"
+
+// SFML
+#include <SFML/Graphics.hpp>
+
+namespace BoBRobotics {
+namespace Viz {
+namespace SFML {
+
+/**!
+ * \brief Drive a robot with an SFML window event
+ *
+ * Returns true if event was handled or false otherwise.
+ */
+template<class RobotType,
+         std::enable_if_t<!Robots::IsAckermann<RobotType>::value, int> = 0>
+bool drive(RobotType &robot, const sf::Event &event)
+{
+    switch (event.type) {
+    case sf::Event::KeyPressed:
+        switch (event.key.code) {
+        case sf::Keyboard::Up:
+            robot.moveForward(1.f);
+            return true;
+        case sf::Keyboard::Down:
+            robot.moveForward(-1.f);
+            return true;
+        case sf::Keyboard::Left:
+            robot.turnOnTheSpot(-0.5f);
+            return true;
+        case sf::Keyboard::Right:
+            robot.turnOnTheSpot(0.5f);
+            return true;
+        default:
+            return false;
+        }
+        break;
+    case sf::Event::KeyReleased:
+        switch (event.key.code) {
+        case sf::Keyboard::Up:
+        case sf::Keyboard::Down:
+        case sf::Keyboard::Left:
+        case sf::Keyboard::Right:
+            robot.stopMoving();
+            return true;
+        default:
+            return false;
+        }
+    default:
+        return false;
+    }
+}
+
+/**!
+ * \brief Drive an Ackermann-type robot with an SFML window event
+ *
+ * The types of maxSpeed and maxTurn are templatable so that one can pass in a
+ * speed in e.g. m/s. Defaults to using maximum speed and turn.
+ *
+ * Returns true if event was handled or false otherwise.
+ */
+template<class RobotType, class SpeedType = float, class TurnType = float,
+         std::enable_if_t<Robots::IsAckermann<RobotType>::value, int> = 0>
+bool
+drive(RobotType &robot, const sf::Event &event, SpeedType maxSpeed = 1.f,
+      TurnType maxTurn = 1.f)
+{
+    /*
+     * We need to keep track of these values between calls. This is a slightly
+     * gross way of doing things, but making a templatable RobotControl class
+     * to manage state instead seems grosser.
+     */
+    static SpeedType velocity{};
+    static TurnType turn{};
```

Comment:
What's the purpose of this? It seems a really bad idea as the values will be shared between multiple robots e.g. in a simulation. Why not just cache the velocity and turn values in the robot class (like tank does)? They're also useful for dead reckoning. Also, why doesn't the Ackermann robot just keep doing whatever the last command was if you don't call drive like all the other robot types?

Reply:
I actually removed the value-caching from ``TankBase`` as I felt that having the extra state in there was making things more complicated (i.e. all subclasses then had to make sure that they told ``TankBase`` when they were updating the motor commands).

The Ackermann robot *will* carry on doing whatever the last command was (``velocity`` and ``turn`` are only zero-initialised in the first call to ``drive()``). I admit that it is a bit gross to have this hidden state lurking in ``drive()`` *however* I don't really want to make an ``AckermannBase`` class which carries around this state either :shrug:. (I suppose the other option would be to make a templated class purely for the robot control.) The "multiple robots" problem is *really* unlikely to be an issue because we're never going to be controlling multiple robots with this interface anyway.